<a href="https://colab.research.google.com/github/zamoralinokevin-eng/Maestria-en-IA/blob/main/Actividad_Calificada_07_Datos_Duplicados_271753.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 2026-09-24 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** Kevin Josue Zamora Lino | |
| **Matrícula** 271753 | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [7]:
#Duplicado de Hoteles con keep='first'
duplicados_first = hoteles.duplicated(keep='first').sum()
print("Filas duplicadas con keep='first':",duplicados_first )
#Duplicado de Hoteles con keep='last'
duplicados_last = hoteles.duplicated(keep='last').sum()
print("Filas duplicadas con keep='last':", duplicados_last)
#Duplicado de Hoteles con keep=False
duplicados_false = hoteles.duplicated(keep=False).sum()
print("Filas duplicadas con keep=False:", duplicados_false)

Filas duplicadas con keep='first': 31994
Filas duplicadas con keep='last': 31994
Filas duplicadas con keep=False: 40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [58]:
#Calculo del porcentaje de filas sobrantes
porcentaje_filas_sobrantes = duplicados_first / len(hoteles) * 100
print(f"El porcentaje de filas sobrantes es: {porcentaje_filas_sobrantes:.1f}%")

#Calculo del número de grupos repetidos opcion 1
conteo_grupos = hoteles.value_counts()
grupos_repetidos = conteo_grupos[conteo_grupos > 1].count()
print(f"El conteo de filas que aparecen mas de una vez (Opcion 1) son: {grupos_repetidos}")

#Calculo del número de grupos repetidos opcion 2
repetidos_separados = hoteles.duplicated(keep='first').sum()
repetidos_juntos = hoteles.duplicated(keep=False).sum()
grupos_opcion2 = repetidos_juntos - repetidos_separados
print(f"El conteo de filas que aparecen mas de una vez (Opcion 2) son: {grupos_opcion2}")




El porcentaje de filas sobrantes es: 26.8%
El conteo de filas que aparecen mas de una vez (Opcion 1) son: 17
El conteo de filas que aparecen mas de una vez (Opcion 2) son: 8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [59]:
#Se crea un copia de hoteles
hoteles_duplicados = hoteles.copy()

#se agrega la columna booleana "es_duplicado" con duplicated()
hoteles_duplicados['es_duplicado'] = hoteles_duplicados.duplicated(keep='first')

#se agrupa por columna "hotel", se muestran los valores duplicados y totales en columnas independientes
hoteles_agrupados = hoteles_duplicados.groupby('hotel')['es_duplicado'].agg(duplicados='sum',total='count').reset_index()

#se calcula el porcentaje con duplicados en el grupo/total del grupo
hoteles_agrupados['porcentaje'] = hoteles_agrupados['duplicados'] / hoteles_agrupados['total'] * 100
hoteles_agrupados


,hotel,duplicados,total,porcentaje
0,City Hotel,25902,79330,32.650952
1,Resort Hotel,6092,40060,15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

_Escribe aquí tu respuesta._ El problema de duplicados en este dataset es significativo y grave, ya que en ambos hoteles, las filas representan un porcentaje de duplicados mayor al 10%. Afecta diferente a ambos hoteles, en especial a City Hotel ya que este presenta un porcentaje de duplicados de 32.65%. El dataset requiere un analisis para ver la causa de los duplicados y determinar si al eliminarlos, no afecte la calidad de los datos. La decision de cómo manejar los datos duplicados debe quedar documentada.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [57]:
#se filtran las filas del dataframe con keep=False
hoteles_filtrado = hoteles[hoteles.duplicated(keep=False)]

#se ordenan las filas filtradas con sort_values() para que las copias queden juntas
hoteles_sorted = hoteles_filtrado.sort_values(by=['hotel', 'arrival_date_year', 'arrival_date_month','lead_time'])

#se muestran las primeras 12 filas con las columnas mencionadas
columnas_a_mostrar = ['hotel','arrival_date_year', 'arrival_date_month', 'lead_time', 'arrival_date_day_of_month', 'adults', 'country', 'market_segment', 'deposit_type', 'adr', 'reservation_status']
hoteles_sorted[columnas_a_mostrar].head(12)

#eleigí las columnas de ordenamiento 'hotel', 'arrival_date_year', 'arrival_date_month' y 'lead_time'
#para que se vea primero el hotel, luego el año y mes en orden ascendente y despues el lead_time
#de cada uno de los regitros para hacer la reservacion

,hotel,arrival_date_year,arrival_date_month,lead_time,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40643,City Hotel,2015,August,0,4,2,PRT,Direct,No Deposit,0.0,Canceled
40644,City Hotel,2015,August,0,4,2,PRT,Direct,No Deposit,0.0,Canceled
40772,City Hotel,2015,August,0,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40802,City Hotel,2015,August,0,7,2,PRT,Direct,No Deposit,75.0,Check-Out
40821,City Hotel,2015,August,0,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
40838,City Hotel,2015,August,0,8,2,PRT,Online TA,No Deposit,89.0,Check-Out
41026,City Hotel,2015,August,0,11,2,PRT,Online TA,No Deposit,75.0,Check-Out
41067,City Hotel,2015,August,0,11,1,PRT,Corporate,No Deposit,88.0,Check-Out
41071,City Hotel,2015,August,0,11,2,PRT,Direct,No Deposit,80.0,Check-Out
41073,City Hotel,2015,August,0,11,1,PRT,Corporate,No Deposit,88.0,Check-Out


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [71]:
#se agrupan las columnas, se incluye dropna=False para incluir los Nan en el conteo
tamano = hoteles.value_counts(dropna=False)

#se muestran los 10 tamaños mas frecuentes de los grupos
print(tamano.value_counts().head(10))

#se indica el tamaño del grupo mas grande y su combinacion de valores
tamano_grupo_max = tamano.iloc[0]
print(f"El tamaño del grupo más grande es: {tamano_grupo_max}")
combinacion_valores = tamano.index[0]
print(f"La combinación de valores del grupo más grande es: {combinacion_valores}")

count
1     79225
2      5034
3       969
4       497
5       275
6       171
10      123
7       115
8       103
20       97
Name: count, dtype: int64
El tamaño del grupo más grande es: 180
La combinación de valores del grupo más grande es: ('City Hotel', np.int64(1), np.int64(277), np.int64(2016), 'November', np.int64(46), np.int64(7), np.int64(1), np.int64(2), np.int64(2), np.float64(0.0), np.int64(0), 'BB', 'PRT', 'Groups', 'TA/TO', np.int64(0), np.int64(0), np.int64(0), 'A', 'A', np.int64(0), 'Non Refund', np.float64(nan), np.float64(nan), np.int64(0), 'Transient', np.float64(100.0), np.int64(0), np.int64(0), 'Canceled', '2016-04-04')


### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [85]:
#se crea un dataset con los valores suplicados
filas_duplicadas = hoteles[hoteles.duplicated(keep=False)]

#se compara market_segment en ambos datasets
print("***Comparacion de market_segment***\n")
print("Dataset Completo:\n", hoteles['market_segment'].value_counts(normalize=True).head(3))
print("\nFilas Duplicadas:\n", filas_duplicadas['market_segment'].value_counts(normalize=True).head(3))


#se compara market_segment en ambos datasets
print("\n\n***Comparacion de deposit_type***\n")
print("Dataset Completo:\n", hoteles['deposit_type'].value_counts(normalize=True))
print("\nFilas Duplicadas:\n", filas_duplicadas['deposit_type'].value_counts(normalize=True))

#se muestra tasa de cancelacion
print("\n\n***Tasa de cancelacion is_canceled***\n")
print(f"Dataset Completo: {hoteles['is_canceled'].mean()}")
print(f"Filas Duplicadas: {filas_duplicadas['is_canceled'].mean()}")

***Comparacion de market_segment***

Dataset Completo:
 market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Name: proportion, dtype: float64

Filas Duplicadas:
 market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Name: proportion, dtype: float64


***Comparacion de deposit_type***

Dataset Completo:
 deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64

Filas Duplicadas:
 deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64


***Tasa de cancelacion is_canceled***

Dataset Completo: 0.37041628277075134
Filas Duplicadas: 0.583866550479273


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

_Escribe aquí tu respuesta._
a) Es probable que pueda ser por una combinacion de ambas. Segun investigacion adicional, el hotel puede bloquear un grupo de habitaciones para un evento privado con los mismos datos, por lo que los registros se repiten para considerar todas las reservas de la empresa.

b) Duplicated() de pandas trata los valores NaN como identicos entre si. Si las filas coindicen con los valores en las columnas, incluyendo los Nan, se contarán como duplicadas. Ejemplo, si las reservas e hicieron por agentes distintos pero se quedaron vacias, la funcion las puede llegar a agrupar como si fuese la misma sesgando los datos.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [86]:
#se crea hoteles_limpio eliminando los duplicados. Se renumera el indice
hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True)

#se imprime el nuero de filas antes, despues y cuantas se eliminaron
filas_antes = len(hoteles)
filas_despues = len(hoteles_limpio)
filas_eliminadas = filas_antes - filas_despues
print(f"Filas antes: {filas_antes}")
print(f"Filas despues: {filas_despues}")
print(f"Filas eliminadas: {filas_eliminadas}")


Filas antes: 119390
Filas despues: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [90]:
#se calculan metricas para el hoteles (dataset de inicio)
resumen_original = hoteles.groupby('hotel').agg(
    numero_reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    tarifa_diaria_promedio=('adr', 'mean')
).reset_index()

#se calculan metricas para el hotel sin duplicados
resumen_limpio = hoteles_limpio.groupby('hotel').agg(
    numero_reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    tarifa_diaria_promedio=('adr', 'mean')
).reset_index()

#se guardaron las medias en los dataframes de arriba
print("Resumen original:\n", resumen_original)
print("\nResumen limpio:\n", resumen_limpio)


Resumen original:
           hotel  numero_reservas  tasa_cancelacion  tarifa_diaria_promedio
0    City Hotel            79330          0.417270              105.304465
1  Resort Hotel            40060          0.277634               94.952930

Resumen limpio:
           hotel  numero_reservas  tasa_cancelacion  tarifa_diaria_promedio
0    City Hotel            53428          0.300386              110.985944
1  Resort Hotel            33968          0.234809               99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [92]:
#se añaden las etiquetas de las versiones de los dataframes
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

#se unen los dataframes con pd.concat() en uno solo
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)

#se ordenan los valores por hotel
comparacion = comparacion.sort_values(by='hotel').reset_index(drop=True)
comparacion

,hotel,numero_reservas,tasa_cancelacion,tarifa_diaria_promedio,version
0,City Hotel,79330,0.417270,105.304465,original
1,City Hotel,53428,0.300386,110.985944,sin duplicados
2,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

_Escribe aquí tu respuesta._
a) La tasa de cancelacion es la metrica que cambia mas al eliminar duplicados. En City Hotel la tasa original es de 0.41 y en el modificado es de 0.30. Esta informacion se corrobora con la actividad 2, porque en City Hotel hay muchos duplicados en la parte de Non Refund. AL eliminar los duplicados, el dataframe elimina el peso de estos valores.

b) Lo que primero ocurriria en el entrenamiento es el overfitting o sobreajuste por los registros duplicados. El modelo memorizará el ruido en vez de aprender los patrones importantes. En la evalauacion, se puede producir el data leakage cuando se dividan los datos y los duplicados queden repartidos entre los grupos del dataset de entrenamiento y validacion generando metricas irreales y falsas.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [94]:
#se verifican si existen duplicados exactos
print("Los duplicados que quedan son: ", hoteles_limpio.duplicated().sum())

#se obtienen ambos indices y se verifica la estructura del indice
esperado = list(range(0, len(hoteles_limpio)))
actual = hoteles_limpio.index.tolist()

#se genera una condicinal para la verifiacion del indice
if actual == esperado and not hoteles_limpio.index.duplicated().any():
    print("Verificación exitosa: 0 duplicados exactos y el índice corre limpio de 0 a", len(hoteles_limpio)-1)

Los duplicados que quedan son:  0
Verificación exitosa: 0 duplicados exactos y el índice corre limpio de 0 a 87395


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [97]:
#se identifican todas las columnas excepto reservation_status_date
columnas_analisis_parcial = [col for col in hoteles_limpio.columns if col != 'reservation_status_date']

#se cuentan las filas sobrantes con el criterio establecido
duplicado_sobrantes = hoteles_limpio.duplicated(subset=columnas_analisis_parcial, keep='first').sum()
print(f" Las filas sobrantes detectadas: {duplicado_sobrantes}")

#se extraen y ordenan las filas involucradas
parciales_involucrados = hoteles_limpio[hoteles_limpio.duplicated(subset=columnas_analisis_parcial, keep=False)]
parciales_ordenados = parciales_involucrados.sort_values(by=columnas_analisis_parcial)

#se declaran las columnas establecidas
columnas_parcial_show = [
    'hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
    'arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date'
]
print("\nMuestra de filas con duplicados parciales:")
print(parciales_ordenados[columnas_parcial_show].head())

#se muestra la distribucion de reservation_status en estas filas
print("\n\nDistribución del Estado de Reserva:")
print(parciales_involucrados['reservation_status'].value_counts())


 Las filas sobrantes detectadas: 245

Muestra de filas con duplicados parciales:
            hotel  lead_time  arrival_date_year arrival_date_month  \
35572  City Hotel          9               2015          September   
35574  City Hotel          9               2015          September   
51267  City Hotel         12               2017               July   
51268  City Hotel         12               2017               July   
54170  City Hotel         14               2015            October   

       arrival_date_day_of_month    adr reservation_status  \
35572                         10   95.0           Canceled   
35574                         10   95.0           Canceled   
51267                         15  139.0           Canceled   
51268                         15  139.0           Canceled   
54170                         15  100.0           Canceled   

      reservation_status_date  
35572              2015-09-03  
35574              2015-09-01  
51267              2017-07-10

**Respuesta 4.2:**

_Escribe aquí tu respuesta._ Estas filas podrian representar las reservas identicas grupales realizadas por un mismo cliente pero que parte de su movimiento se registro en fechas diferentes. Al igual que en una pregunta anterior, antes de eliminar, se debe consultar con la empresa si estos datos son relevantes y en caso de eliminarse, debe documentarse el criterio por el cual se eliminó.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

_Escribe aquí tu respuesta._
Se utilizo el duplicado exacto, es decir, que los valores sean los mismos en cada columna de la fila, incluyendo NaN.
Se usó keep='first' en el filtrado porque no existe una marca de tiempo como campo en los datos o cualquier otro indicador tipo ID que ayude a diferenciar entre filas.
Resultados:
filas antes: 119390
filas despues: 87396
filas eliminadas: 31994
Existe el supuesto de que se eliminaron duplicados por errores de captura o de sistema asi como las reservaciones grupales de un mismo cliente. Las limitaciones es que el dataset no contiene una columna tipo ID o timestamp para una mejor identificacion de las filas por lo que existe la posibilidad de que se hayn eliminado registros relevantes para la empresa.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

_Escribe aquí tu respuesta._
Una situacion en la que no conviene eliminar duplicados exactos es si el dataset incluye una columna 'precio' por cada uno de los datos duplicados y estos sean el resultado de una reservacion en grupo. ya que no se calcularia bien el ingreso del hotel por todas las reservaciones del grupo, afectando los resultados finacieros del hotel.
Las columnas adicionales que se pedirian son las relacionandas con el ID de la reservacion, una categoria de si la reservacion es indivdual o grupal, el precio de la reservacion y un timestamp donde se hizo legitima la reservacion.
La mascara booleana de duplicated().sum() == 0 no garantiza el dataset libre de duplicados ya que cuando se avalua con duplicated(), este busca un 100% de coincidencia en las filas, sesgando filas que se repiten solo en ciertos campos o columnas.Si no se evalua bien el dataset, este seguirá teniendo duplicados ocultos a pesar de que esta condicion se cumpla

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.